feat(user): assign profile users photo (#433)

bertapasamontes opened the pull request:
### Description
This PR completes task **433**, part of user story 388 - "As a user, I can see my role and be able to modify it in the profile drop-down menu.".

A GET request was made to retrieve the user's profile picture URL, and it was added to the navbar, replacing the profile icon.
In addition, a refactor of shared styles was carried out.
Reusable style rules were extracted and moved into the _global.scss file to promote consistency and reduce duplication across components.

### How to test it:
- Sign in with github
- Check if your github profile image is displayed in the menu.

### Before and after comparison
**Before:**
![image](https://github.com/user-attachments/assets/ff2a7abd-e20a-4e89-be72-db46fe628fd8)
![image](https://github.com/user-attachments/assets/0752b939-fc56-4901-a1a2-dc6bbb792e61)

**After (desktop):**
![image](https://github.com/user-attachments/assets/4893c4f6-815d-4a69-9b67-ac0da44cfc59)

**After (mobile):**
![image](https://github.com/user-attachments/assets/32f1d6df-94d4-4e49-9201-23d3bb353dd0)

### Files changed:
- environmet
- auth.service
- desktop-nav
- mobile-nav
- _global.scss


### Auto self-check:

- [x] I tested my changes locally and verified they work as expected
- [ ] The PR has a clear and focused purpose (only one feature, fix or refactor)
- [x] Title, description, and changelog (if needed) are filled in correctly
- [x] There are no leftover merge conflicts or unrelated changes from previous branches
- [ ] All automated checks (like SonarCloud) have passed
- [x] I responded to all previous review comments and only resolved those I truly addressed

bertapasamontes committed:
user's profile photo in desktop-nav

bertapasamontes committed:
user's profile photo in mobile-nav + styles updated

bertapasamontes committed:
added class erased by error. dropdown works

bertapasamontes committed:
test passed

basantades commented on src/app/core/layout/header/desktop-nav/desktop-nav.component.scss:
```diff
@@ -99,6 +97,19 @@
     display: inline-block;
   }
 
+  .userWithPhoto {
```
I'd suggest updating the .userWithPhoto class to the following:

.userWithPhoto {
  border-radius: 50%;
  width: 40px;
  height: 40px;
  position: relative;
  display: inline-block;
  padding: 0;
  overflow: hidden;
}


border: unset seems unnecessary here, so I removed it.

Updated the size to 40px to better match the Figma design and avoid the top cropping issue (some of the container/layout CSS is causing unexpected shifts).

Added padding: 0 to prevent a strange white border caused by inherited padding from a parent element.

Added overflow: hidden so the image doesn't need a separate class to apply the border radius. This way, you can control the circular shape from a single place.

Let me know if that works for you!

bertapasamontes replied:
> I'd suggest updating the .userWithPhoto class to the following:
> 
> .userWithPhoto { border-radius: 50%; width: 40px; height: 40px; position: relative; display: inline-block; padding: 0; overflow: hidden; }
> 
> border: unset seems unnecessary here, so I removed it.
> 
> Updated the size to 40px to better match the Figma design and avoid the top cropping issue (some of the container/layout CSS is causing unexpected shifts).
> 
> Added padding: 0 to prevent a strange white border caused by inherited padding from a parent element.
> 
> Added overflow: hidden so the image doesn't need a separate class to apply the border radius. This way, you can control the circular shape from a single place.
> 
> Let me know if that works for you!

done :)

basantades commented on src/app/core/layout/header/mobile-nav/mobile-nav.component.scss:
```diff
@@ -109,6 +109,19 @@ $navbar-toggler-transition: none;
       display: inline-block;
     }
+
+    .userWithPhoto {
```
Same as in the other CSS

I'd suggest updating the .userWithPhoto class to the following:

.userWithPhoto {
border-radius: 50%;
width: 40px;
height: 40px;
position: relative;
display: inline-block;
padding: 0;
overflow: hidden;
}

border: unset seems unnecessary here, so I removed it.

Updated the size to 40px to better match the Figma design and avoid the top cropping issue (some of the container/layout CSS is causing unexpected shifts).

Added padding: 0 to prevent a strange white border caused by inherited padding from a parent element.

Added overflow: hidden so the image doesn't need a separate class to apply the border radius. This way, you can control the circular shape from a single place.

Let me know if that works for you!

bertapasamontes replied:
> Same as in the other CSS
> 
> I'd suggest updating the .userWithPhoto class to the following:
> 
> .userWithPhoto { border-radius: 50%; width: 40px; height: 40px; position: relative; display: inline-block; padding: 0; overflow: hidden; }
> 
> border: unset seems unnecessary here, so I removed it.
> 
> Updated the size to 40px to better match the Figma design and avoid the top cropping issue (some of the container/layout CSS is causing unexpected shifts).
> 
> Added padding: 0 to prevent a strange white border caused by inherited padding from a parent element.
> 
> Added overflow: hidden so the image doesn't need a separate class to apply the border radius. This way, you can control the circular shape from a single place.
> 
> Let me know if that works for you!

done :)

basantades commented on src/app/core/layout/header/desktop-nav/desktop-nav.component.html:
```diff
@@ -13,9 +13,16 @@
-        <img src="assets/img/icon/user.svg" alt="User" style="border: none; padding: 0; background: none;">
-      </button>
+       @if (!userPhoto) {
+        <button class="user" (click)="toggleDropdown()">
```
Please consider adding aria-label="User menu" to the <button> element to improve accessibility for screen readers.

bertapasamontes replied:
> Please consider adding aria-label="User menu" to the element to improve accessibility for screen readers.

done :)

basantades commented on src/app/core/layout/header/desktop-nav/desktop-nav.component.html:
```diff
@@ -13,9 +13,16 @@
+        </button>
+      }
+      @else {
+        <button class="user userWithPhoto" (click)="toggleDropdown()">
```
Please consider adding aria-label="User menu" to the <button> element to improve accessibility for screen readers.

bertapasamontes replied:
> Please consider adding aria-label="User menu" to the element to improve accessibility for screen readers.

done :)

basantades commented on src/app/core/layout/header/mobile-nav/mobile-nav.component.html:
```diff
@@ -24,9 +24,16 @@
-            <img src="assets/img/icon/user.svg" alt="User" style="border: none; padding: 0; background: none;">
-          </button>
+          @if (!userPhoto) {
+            <button class="user-mobile" (click)="toggleDropdown()">
```
Please consider adding aria-label="User menu" to the <button> element to improve accessibility for screen readers.

bertapasamontes replied:
> Please consider adding aria-label="User menu" to the element to improve accessibility for screen readers.

done :)

basantades commented on src/app/core/layout/header/mobile-nav/mobile-nav.component.html:
```diff
@@ -24,9 +24,16 @@
+            </button>
+          }
+          @else {
+            <button class="user-mobile userWithPhoto" (click)="toggleDropdown()">
```
Please consider adding aria-label="User menu" to the <button> element to improve accessibility for screen readers.

bertapasamontes replied:
done :)

basantades commented:
Hi! I added a few inline suggestions related to the CSS and accessibility, such as simplifying the avatar styling and improving screen reader support with aria-label.

Here's a quick before/after comparison to show the visual effect of the CSS adjustments:
![image](https://github.com/user-attachments/assets/86408e53-6f54-41b0-8c20-61101ae2888d)
![image](https://github.com/user-attachments/assets/393b80bf-8101-4aef-bb8f-917de7c0a0d7)

Additionally, it might be worth considering extracting the user profile photo into a separate component (e.g. UserPhotoComponent) since it's used in multiple places and could help keep the layout components cleaner and more focused.

Let me know what you think!

bertapasamontes committed:
aria label added + userWithPhoto styles updated

bertapasamontes commented:
> Hi! I added a few inline suggestions related to the CSS and accessibility, such as simplifying the avatar styling and improving screen reader support with aria-label.
> 
> Here's a quick before/after comparison to show the visual effect of the CSS adjustments: ![image](https://private-user-images.githubusercontent.com/144531213/447243218-86408e53-6f54-41b0-8c20-61101ae2888d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDgyNTcwMjcsIm5iZiI6MTc0ODI1NjcyNywicGF0aCI6Ii8xNDQ1MzEyMTMvNDQ3MjQzMjE4LTg2NDA4ZTUzLTZmNTQtNDFiMC04YzIwLTYxMTAxYWUyODg4ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyNlQxMDUyMDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MmQ5NWQ3OTNkODlhMWNmNTcwNjg3ZjIxNzJkZTczZTkwZGVhZDM5ZTNhOTQyMDIyMDk1ZWI2YTk5ZWYyZmI1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.l4imnBj1Pf4uFek4I-VUNtbUPbhJ4ssHk4oyuydTDMc) ![image](https://private-user-images.githubusercontent.com/144531213/447243251-393b80bf-8101-4aef-bb8f-917de7c0a0d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDgyNTcwMjcsIm5iZiI6MTc0ODI1NjcyNywicGF0aCI6Ii8xNDQ1MzEyMTMvNDQ3MjQzMjUxLTM5M2I4MGJmLTgxMDEtNGFlZi1iYjhmLTkxN2RlN2MwYTBkNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyNlQxMDUyMDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iOTU4YzY5ZTk2NDM3MjdhZWYzMzk3MWZkMDI5Y2RhYTYyZDA4MzZiMzM2YmUzNGI0YTkzZGFiZDJhOTY4YmYwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Vd-H16w_oodX48vAsTebwuWMpnxiNKZf43-oEiLsGX0)
> 
> Additionally, it might be worth considering extracting the user profile photo into a separate component (e.g. UserPhotoComponent) since it's used in multiple places and could help keep the layout components cleaner and more focused.
> 
> Let me know what you think!

Thanks, Alberto, for your suggestions. I've already implemented the style changes. Regarding the new component, I believe it would be more efficient to refactor the entire menu into a new component, rather than creating one solely for the user's photo. What do you think?

basantades commented:
> Thanks, Alberto, for your suggestions. I've already implemented the style changes. Regarding the new component, I believe it would be more efficient to refactor the entire menu into a new component, rather than creating one solely for the user's photo. What do you think?

Yes, that sounds like a good idea. Still, even when refactoring the menu, it might be a good idea to call a small avatar component from within it 🙂.

bertapasamontes committed:
style(menu): improve mobile layout and refactor styles to _global.scss

bertapasamontes committed:
style(user's photo): adjust overflow

bertapasamontes committed:
refactor(html): remove unused CSS class from desktop and mobile html files

basantades reviewed (approved):
All good on my side.

ivilarop commented on src/app/core/layout/header/desktop-nav/desktop-nav.component.html:
```diff
@@ -13,9 +13,16 @@
-      </button>
+       @if (!userPhoto) {
+        <button class="user" (click)="toggleDropdown()">
+          <img src="assets/img/icon/user.svg" alt="User" style="border: none; padding: 0; background: none;">
```
The image's alt attribute is statically set to "User." To maintain consistency with the application's internationalization system and improve accessibility, would it be advisable to translate it?

bertapasamontes replied:
> The image's alt attribute is statically set to "User." To maintain consistency with the application's internationalization system and improve accessibility, would it be advisable to translate it?

done

ivilarop commented on src/app/core/layout/header/desktop-nav/desktop-nav.component.html:
```diff
@@ -13,9 +13,16 @@
+      }
+      @else {
+        <button class="user userWithPhoto" (click)="toggleDropdown()">
+          <img [src]="userPhoto" alt="User" class="img-fluid" aria-label="User menu">
```
The image's alt attribute is statically set to "User." To maintain consistency with the application's internationalization system and improve accessibility, would it be advisable to translate it?

bertapasamontes replied:
> The image's alt attribute is statically set to "User." To maintain consistency with the application's internationalization system and improve accessibility, would it be advisable to translate it?

done

ivilarop commented on src/app/core/layout/header/desktop-nav/desktop-nav.component.html:
```diff
@@ -13,9 +13,16 @@
-      <button class="user" (click)="toggleDropdown()">
-        <img src="assets/img/icon/user.svg" alt="User" style="border: none; padding: 0; background: none;">
-      </button>
+       @if (!userPhoto) {
```
The if/else blocks are similar. Both buttons do the same thing (toggleDropdown()), the only thing that changes is: The class (user vs user userWithPhoto) and The content (different img). Recommendation: extract the dynamic class and conditional content with ngClass

bertapasamontes replied:
> The if/else blocks are similar. Both buttons do the same thing (toggleDropdown()), the only thing that changes is: The class (user vs user userWithPhoto) and The content (different img). Recommendation: extract the dynamic class and conditional content with ngClass

done

ivilarop commented on src/app/core/layout/header/desktop-nav/desktop-nav.component.ts:
```diff
@@ -32,8 +33,9 @@ export class DesktopNavComponent implements OnInit, OnDestroy{
-    this._authService.getUserRole().subscribe((userRole) => {
-      this.currentRole = userRole
+
+    this._authService.getUserPhoto().subscribe((userPhoto) => {
```
Tip: Since each call (getUsername, getUserPhoto, updateUserRoleAndUserNameFromToken) is handled differently and there's no way to unify it in a single endpoint, it's justified to maintain separate subscriptions. Still, we could improve readability by extracting that logic into a loadUserBasicData()-like method.

This helps keep ngOnInit() clean and separate by responsibility, especially if we add error handling or loading in the future.

bertapasamontes replied:
> Tip: Since each call (getUsername, getUserPhoto, updateUserRoleAndUserNameFromToken) is handled differently and there's no way to unify it in a single endpoint, it's justified to maintain separate subscriptions. Still, we could improve readability by extracting that logic into a loadUserBasicData()-like method.
> 
> This helps keep ngOnInit() clean and separate by responsibility, especially if we add error handling or loading in the future.

done

ivilarop commented on src/app/core/layout/header/mobile-nav/mobile-nav.component.html:
```diff
@@ -24,9 +24,16 @@
-          <button class="user-mobile" (click)="toggleDropdown()">
-            <img src="assets/img/icon/user.svg" alt="User" style="border: none; padding: 0; background: none;">
-          </button>
+          @if (!userPhoto) {
```
same comments that desktop

bertapasamontes replied:
> same comments that desktop

done

ivilarop commented on src/app/core/layout/header/mobile-nav/mobile-nav.component.ts:
```diff
@@ -31,8 +31,8 @@ export class MobileNavComponent implements OnInit, OnDestroy{
 
-    this._authService.getUserRole().subscribe((userRole) => {
-      this.currentRole = userRole
+    this._authService.getUserPhoto().subscribe((userPhoto) => {
```
same comments that desktop

bertapasamontes replied:
> same comments that desktop

done

ivilarop commented on src/app/services/auth.service.spec.ts:
```diff
@@ -164,4 +169,22 @@ describe('AuthService', () => {
+    const photoUrl = 'https://github.com/avatar.jpg'
+    const mockResponse = { avatar_url: photoUrl };
+
+    // Asigna un username al servicio usando spy o cambia visibilidad de la propiedad si necesario
```
necessary?

bertapasamontes replied:
> necessary?

done

ivilarop commented on src/app/services/auth.service.ts:
```diff
@@ -51,6 +70,9 @@ export class AuthService {
       this.usernameSubject.next(this.username);
       const userId = decodedToken?.uuid ?? null;
       this.userIdSubject.next(userId);
+      this.getUserPhoto().subscribe(photo => {
```
Suggestion: Subscribing to getUserPhoto() inside updateUserRoleAndUserNameFromToken() seems unnecessary, since the getUserPhoto() method itself internally assigns the value to this.userPhoto. We could remove this subscription, since nothing useful is being done with the received value. If we want to expose this data to other components in the future, we could also emit it through a Subject or BehaviorSubject.

bertapasamontes replied:
> Suggestion: Subscribing to getUserPhoto() inside updateUserRoleAndUserNameFromToken() seems unnecessary, since the getUserPhoto() method itself internally assigns the value to this.userPhoto. We could remove this subscription, since nothing useful is being done with the received value. If we want to expose this data to other components in the future, we could also emit it through a Subject or BehaviorSubject.

done

ivilarop commented on src/app/services/auth.service.ts:
```diff
@@ -22,12 +22,31 @@ export class AuthService {
     this.translate.use('ca')
   }
 
+  getUserPhoto(): Observable<string> {
```
Currently, a BehaviorSubject is used, and a manual subscribe() is performed within the service itself. Although functional, this is not the most idiomatic way to work with RxJS.

Additionally, the 'error' string is emitted in the event of a failure, which can mix types in the observable (URL vs. 'error'), making subsequent handling from the component difficult.

It would be interesting to investigate how to improve this pattern by using RxJS operators such as pipe, map, tap, and catchError, while maintaining the declarative flow and avoiding subscribing directly from the service.

bertapasamontes replied:
> Currently, a BehaviorSubject is used, and a manual subscribe() is performed within the service itself. Although functional, this is not the most idiomatic way to work with RxJS.
> 
> Additionally, the 'error' string is emitted in the event of a failure, which can mix types in the observable (URL vs. 'error'), making subsequent handling from the component difficult.
> 
> It would be interesting to investigate how to improve this pattern by using RxJS operators such as pipe, map, tap, and catchError, while maintaining the declarative flow and avoiding subscribing directly from the service.

done

ivilarop commented on src/environments/environment.ts:
and shouldn't it be added to the environment.prod to work when deployed to development?

bertapasamontes replied:
> and shouldn't it be added to the environment.prod to work when deployed to development?

done

bertapasamontes committed:
feat(environment.prod): added variable

bertapasamontes committed:
feat(changelog + header-nav): updated changelog + refactor nav

bertapasamontes committed:
refact(desktop): deleted commented lines

bertapasamontes committed:
feat(i18n): added new lines to traduce alt User's photo

bertapasamontes committed:
refactor(desktop html): updated classes

bertapasamontes committed:
refactor(mobile html): updated html to be as like as desktop, refactor(mobile ts): updated toggle to work. rebase went wrong and deleted important code

bertapasamontes committed:
refactor(auth): update getUserPhoto with RxJS pipe and clean up auth.service.spec

Refactored getUserPhoto to use a declarative RxJS pipeline with pipe, filter, switchMap, map, tap, and catchError. This avoids manual subscriptions and mixed-type emissions, improving code clarity and maintainability.

Also removed unused and unnecessary lines from auth.service.spec to simplify the test file and ensure alignment with the updated service logic.

bertapasamontes committed:
refactor(dektop): deleted unused import

bertapasamontes committed:
test(desktop-nav, mobile-nav and authservice): updated files to pass tests
